backend/lib/src/services/auth.rs

```diff
@@ -1,73 +1,589 @@
+    ///
+    /// This follows the EIP-4361 standard for Sign-In with Ethereum messages.
+    /// The message format ensures compatibility with wallet signing interfaces
+    /// and provides a standardized authentication flow.
+    fn construct_auth_message(address: &str, domain: &str, nonce: &str, chain_id: u64) -> String {
+        let scheme = "https";
+
+        // TODO: make uri match endpoint
+        let uri = format!("{scheme}://{domain}/auth/nonce");
+        let statement = "I authenticate with this MSP Backend with my address";
+        let version = 1;
+        let issued_at = chrono::Utc::now().to_rfc3339();
+
+        // SIWE message format as per EIP-4361
+        format!(
+            "{scheme}://{domain} wants you to sign in with your Ethereum account:\n\
+            {address}\n\
+            \n\
+            {statement}\n\
+            \n\
+            URI: {uri}\n\
+            Version: {version}\n\
+            Chain ID: {chain_id}\n\
+            Nonce: {nonce}\n\
+            Issued At: {issued_at}",
+        )
+    }
+
+    /// Generate a JWT token for the given address
+    fn generate_jwt(&self, address: &str) -> Result<String, Error> {
+        let now = Utc::now();
+        let exp = now + JWT_EXPIRY_OFFSET;
+
+        let claims = JwtClaims {
+            address: address.to_string(),
+            exp: exp.timestamp(),
+            iat: now.timestamp(),
+        };
+
+        jsonwebtoken::encode(&Header::default(), &claims, &self.encoding_key)
+            .map_err(|_| Error::Internal)
+    }
+
+    /// Generate a SIWE-compliant message for the user to sign
+    ///
+    /// The message will expire after a given time
+    pub async fn challenge(&self, address: &str, chain_id: u64) -> Result<NonceResponse, Error> {
+        // Validate address before generating message or storing in cache
         validate_eth_address(address)?;
 
-        Ok(NonceResponse {
-            message: format!(
-                "example.com wants you to sign in with your Ethereum account:\n{}\n\n\
-                Sign in to access your account.\n\n\
-                URI: https://example.com\n\
-                Version: 1\n\
-                Chain ID: {}\n\
-                Nonce: aBcDeF12345\n\
-                Issued At: 2024-01-01T00:00:00Z",
-                address, chain_id
-            ),
-            nonce: "aBcDeF12345".to_string(),
-        })
+        let nonce = Self::generate_random_nonce();
+        let message = Self::construct_auth_message(address, AUTH_SIWE_DOMAIN, &nonce, chain_id);
+
+        // Store message paired with address in storage
+        // Using message as key and address as value
+        self.storage
+            .store_nonce(
+                message.clone(),
+                address.to_string(),
+                AUTH_NONCE_EXPIRATION_SECONDS,
+            )
+            .await
+            .map_err(|_| Error::Internal)?;
+
+        Ok(NonceResponse { message })
     }
```

Comment:
Shouldnt this be based on address? That way we can check if a given address already an active nonce